### PR TITLE
megabatch: env-gated NaN capture wrapper around newton_schulz_func

### PR DESCRIPTION
**Pull request by JohnLangford**
## Summary

Re: #76. The reporter and maintainers cannot consistently reproduce the
NorMuon + gram-newton-schulz + quack-kernels 0.4.1 NaN bug, so the next
debugging step is to capture the actual NS input that triggers the failure
on the affected hardware (6x Blackwell RTX PRO 6000 + DDP).

This PR wraps ``newton_schulz_func`` inside ``megabatch_orthogonalize_async``
with an env-gated capture wrapper. When enabled, each rank that observes a
non-finite NS output writes its own .pt dump (input + output + rank +
shape + epsilon) and by default raises ``RuntimeError``. The check is
**local per rank — no extra collective** — and it short-circuits at one
``os.environ`` lookup when disabled.

## Usage

```bash
DION_NAN_CAPTURE=1 \
DION_NAN_CAPTURE_DIR=/some/scratch/dir \
DION_NAN_CAPTURE_RAISE=1 \  # default; set to 0 to keep training
torchrun ... train.py
```

Per-rank filenames are ``dion_nan_capture_rank{rank}_shape{...}_pid{...}_{ts_ms}.pt``
so concurrent writes from multiple ranks on a shared filesystem do not collide.

## Notes

- Disabled by default. Cost when off: one ``os.environ`` lookup per call.
- The input is ``.detach().clone()``-d up front so we still have the
  offending tensor even if a kernel mutates X in-place.
- Two new module-level imports (``os``, ``time``).

## Test plan

- [x] ``test_capture_disabled_by_default`` — no env, no dumps even when NS returns NaN
- [x] ``test_capture_dumps_on_non_finite_output`` — dump contains correct input/output/shape/rank/epsilon
- [x] ``test_capture_raises_by_default_when_enabled`` — ``RuntimeError`` after dump
- [x] ``test_capture_no_dump_when_finite`` — finite output -> no dump even with env on
- [x] ``test_capture_filename_includes_rank`` — non-zero rank reflected in filename

All five pass single-rank without GPU/NCCL.

### CODE REVIEW

**Comment by donglixp** [comment]
https://github.com/microsoft/dion/pull/102
This PR might be related to the above issue.

**Comment by JohnLangford** [comment]
Closing for now---it may be addressed.
